pylearner.ts

```diff
@@ -1350,7 +1350,119 @@ class ArrayTypeExpression extends TypeExpression {
   }
 }
 
+declare var Immutable: typeof import('./node_modules/immutable/dist/immutable');
+const ImmutableSet = Immutable.Set;
+
+/**
+ * An instance of this class represents a mathematical relation between variable names, the may-alias-relation. The relation is reflexive, symmetric and not transitive. The relation relates variables x and y if they are potential aliases of each other.
+ * 
+ * @author Arthur Spillebeen
+ */
+class MayAliasRelation {
+  /**  Set of may-alias-sets where a may-alias-set is a collection of possible aliases. If two variables are in one of the may-alias-sets together, these variables are possible aliases.
+   *
+   * @invar a may-alias-set is never a subset of another may-aliast-set
+   * @invar a may-alias-set consists of at least two variables
+   */
+   private readonly mayAliasSets: Immutable.Set<Immutable.Set<string>>;
+
+  /**
+   * Constructor to create may-alias-relation that relates variable names if and only if variables are potential aliases by given links.
+   * 
+   * @param mayAliasSets may-alias-sets to be initialized with
+   */
+  constructor(mayAliasSets?: Immutable.Set<Immutable.Set<string>>) {
+    if (mayAliasSets) {
+      this.mayAliasSets = mayAliasSets.filter(s => s.size > 1);
+      this.mayAliasSets = this.removeSubsets();
+    } else 
+      this.mayAliasSets = ImmutableSet();
+  }
+
+  /**
+   * Returns a relation that relates variables x and y if and only if `this` holds for x and y and both x and y are not elements of `variables`.
+   * 
+   * @param variables variables to remove all links to
+   * @returns  relation with links to variables removed
+   */
+  removeVariables(variables: Immutable.Set<string>): MayAliasRelation {
+    let newMayAliasSets = this.mayAliasSets.map(s => s.subtract(s.intersect(variables)));
```

Comment:
```suggestion
    let newMayAliasSets = this.mayAliasSets.map(s => s.subtract(variables));
```
(My earlier suggestion to use `intersect` here was wrong; sorry!)